sockeye/inference.py

```diff
@@ -802,42 +848,51 @@ def _beam_search(self,
         # Maximum output length
         max_output_length = self.models[0].get_max_output_length(source_length)
 
-        # sequences: (beam_size, output_length), pre-filled with <s> symbols on index 0
-        sequences = mx.nd.array(np.full((self.beam_size, max_output_length), C.PAD_ID), dtype='int32', ctx=self.context)
+        # General data structure: each row has batch_size * beam blocks for the 1st sentence, with a full beam,
+        # then the next block for the 2nd sentence and so on
+
+        # sequences: (batch_size * beam_size, output_length), pre-filled with <s> symbols on index 0
+        sequences = mx.nd.full((self.batch_size * self.beam_size, max_output_length), val=C.PAD_ID, ctx=self.context, dtype='int32')
         sequences[:, 0] = self.start_id
 
-        lengths = mx.nd.ones((self.beam_size, 1), ctx=self.context)
-        finished = mx.nd.zeros((self.beam_size,), dtype='int32', ctx=self.context)
+        lengths = mx.nd.ones((self.batch_size * self.beam_size, 1), ctx=self.context)
+        finished = mx.nd.zeros((self.batch_size * self.beam_size,), ctx=self.context, dtype='int32')
 
         # attentions: (beam_size, output_length, encoded_source_length)
-        attentions = mx.nd.zeros((self.beam_size, max_output_length, encoded_source_length), ctx=self.context)
+        attentions = mx.nd.zeros((self.batch_size * self.beam_size, max_output_length, encoded_source_length), ctx=self.context)
 
         # best_hyp_indices: row indices of smallest scores (ascending).
-        best_hyp_indices = mx.nd.zeros((self.beam_size,), ctx=self.context)
+        best_hyp_indices = mx.nd.zeros((self.batch_size * self.beam_size,), ctx=self.context, dtype='int32')
         # best_word_indices: column indices of smallest scores (ascending).
-        best_word_indices = mx.nd.zeros((self.beam_size,), ctx=self.context, dtype='int32')
+        best_word_indices = mx.nd.zeros((self.batch_size * self.beam_size,), ctx=self.context, dtype='int32')
         # scores_accumulated: chosen smallest scores in scores (ascending).
-        scores_accumulated = mx.nd.zeros((self.beam_size, 1), ctx=self.context)
+        scores_accumulated = mx.nd.zeros((self.batch_size * self.beam_size, 1), ctx=self.context)
+
+        best_hyp_indices_np = np.empty((self.batch_size * self.beam_size,), dtype='int32')
+        best_word_indices_np = np.empty((self.batch_size * self.beam_size,), dtype='int32')
+        scores_accumulated_np = np.empty((self.batch_size * self.beam_size,))
 
         # reset all padding distribution cells to np.inf
         self.pad_dist[:] = np.inf
 
-        # (0) encode source sentence
+        # (0) encode source sentence, returns a list
+        # [encoder representation of the source, source_length, initial hidden state of decoder RNN, initial decoder states] tiled to beam size
```

Comment:
This comment is specifically for the RNN, but depending on the model there might be different hidden states. In general the list will contain whatever hidden states the decoder provides.

Reply:
Removed the comment, copied it indeed from the RNN decoder for my notes